// @snap/camera-kit/lib/metrics/operational/Metric.js
const nameDelimiter = "_";
const dimensionDelimiter = ".";
const delimiterRegex = new RegExp(`^${nameDelimiter}+|${nameDelimiter}+$`, "g");
/** @internal */
export class Metric {
    constructor(name, dimensions = {}) {
        this.name = name;
        this.dimensions = dimensions;
    }
    child(constructor, name, dimensions = {}) {
        return new constructor(`${this.name}${nameDelimiter}${name}`, dimensions);
    }
}
/** @internal */
export function joinMetricNames(names) {
    return names.join(nameDelimiter).replace(delimiterRegex, "");
}
/** @internal */
export function serializeMetricDimensions(dimensions) {
    if (Object.keys(dimensions).length === 0)
        return "";
    // The order in which we'll iterate over the dimension keys is governed by the (rather complex) rules for object
    // key iteration in JS -- since ES2015, for string keys of own properties, this is insertion order.
    return `${dimensionDelimiter}${Array.from(Object.entries(dimensions))
        .map((d) => d.join(dimensionDelimiter))
        .join(dimensionDelimiter)}`;
}
